staging/src/k8s.io/client-go/tools/leaderelection/leaderelection.go

```diff
@@ -318,15 +320,38 @@ func (le *LeaderElector) release() bool {
 // on success else returns false.
 func (le *LeaderElector) tryAcquireOrRenew(ctx context.Context) bool {
 	now := metav1.NewTime(le.clock.Now())
+	// 1. fast path to update optimistically assuming that the record observed
+	// last time is the current version.
+	oldObservedRecord := le.getObservedRecord()
+	if le.IsLeader() && le.observedTime.Add(time.Second*time.Duration(oldObservedRecord.LeaseDurationSeconds)).After(now.Time) {
+		oldObservedRecord.RenewTime = now
+		if err := le.config.Lock.Update(ctx, oldObservedRecord); err != nil {
+			klog.Errorf("Failed to update lock optimitically: %v, falling back to slow path", err)
+		} else {
+			le.setObservedRecord(&oldObservedRecord)
+			return true
+		}
+	}
+
 	leaderElectionRecord := rl.LeaderElectionRecord{
 		HolderIdentity:       le.config.Lock.Identity(),
 		LeaseDurationSeconds: int(le.config.LeaseDuration / time.Second),
 		RenewTime:            now,
 		AcquireTime:          now,
 	}
 
-	// 1. obtain or create the ElectionRecord
-	oldLeaderElectionRecord, oldLeaderElectionRawRecord, err := le.config.Lock.Get(ctx)
+	// 2. obtain or create the ElectionRecord
+	var (
+		oldLeaderElectionRecord    *resourcelock.LeaderElectionRecord
+		oldLeaderElectionRawRecord []byte
+		err                        error
+	)
+	// Losing the leader lock is costly so let's not risky it using cache.
+	if le.IsLeader() {
+		oldLeaderElectionRecord, oldLeaderElectionRawRecord, err = le.config.Lock.Get(ctx)
+	} else {
+		oldLeaderElectionRecord, oldLeaderElectionRawRecord, err = le.config.Lock.GetFromCache(ctx)
```

Comment:
I personally against it - with the fast path added above [of reusing the previously written object], this slow path should be exercised extremely rarely.

In which case I basically would switch to:
- start simple [just always read from etcd here, given it's just a fallback]
- add metric to show when the "slow path" is exercised
- add further optimizations only if the metrics from production environments would prove that the slow path is actually exercised frequently enough

Reply:
I agree with what you said, but this is not for leaders, instead it's for the non-leaders to get current lock which has no fast path. The `observedRecord` is updated after this code.

Reply:
So can we divide this PR into two then:
- optimize for leader [as mentioned above]
- a follow-up that would further optimize for non-leader
?

Reply:
Sure, they are in separate commits anyway.